src/lib.rs

```diff
@@ -1,6 +1,560 @@
+use core::panic;
+
 #[doc = include_str!("../README.md")]
 
-/// Say hi to the entire world :)
-pub fn greet() {
-	println!("Hello World!");
+#[derive(PartialEq, Eq)]
+#[derive(Debug)]
+pub enum StringDiffOpKind {
+    Substitute(char, char),  
+    Insert(char),
+    Delete(char),
+    Transpose(usize, usize),
+}
+
+#[derive(Debug)]
+pub struct StringDiffOp {
+    pub kind: StringDiffOpKind,
+    pub index: isize,
+}
+
+pub (crate) fn remove(start: usize, stop: usize, s: &str) -> String {
+    let mut rslt = "".to_string();
+    for (i, c) in s.chars().enumerate() {
+        if start > i || stop < i + 1 {
+            rslt.push(c);
+        }
+    }
+    rslt
+}
+
+pub fn applyDiff(s: &str, diffs: Vec<StringDiffOp>) -> String {
+    let mut new_string : String = s.into();
+    
+    for i in diffs.iter(){
+        match i.kind {
+            StringDiffOpKind::Delete(x) => {
+                new_string = remove(
+                        i.index as usize, 
+                    (i.index as usize)+1 as usize,
+                     &new_string);
+
+            },
+            StringDiffOpKind::Insert(x) => {
+                new_string.push(x);
+            },
+            StringDiffOpKind::Substitute(x,y) => {
+                new_string.replace_range(
+                    (i.index as usize)..( (i.index as usize)+1 as usize),
+                &y.to_string())
+            },
+            _ => {
+                panic!("Unrecoginzed Opperation")
+            }
+        }
+    }
+    return new_string.clone();
+}
+
+pub trait StringDiffAlgorithm {
+    //fn diff<'a>(s1: &'a str, s2: &'a str) -> Vec<StringDiffOp>;
+    fn diff<'a>(&self, s1: &'a str, s2: &'a str) -> Vec<StringDiffOp>;
+    fn distance<'a>(&self, s1: &'a str, s2: &'a str) -> usize;
+}
+
+pub struct HammingDistance {}
+impl HammingDistance{}
+impl StringDiffAlgorithm for HammingDistance {
+    fn diff<'a>(&self, s1: &'a str, s2: &'a str) -> Vec<StringDiffOp>{
+        let mut opp_vec : Vec<StringDiffOp> = Vec::new();
+        if s1.len() != s2.len(){
+            panic!("Strings must be same length");
+        }
+        else{
+            for i in 0..s1.len(){
+                if s1.chars().nth(i).unwrap() != s2.chars().nth(i).unwrap(){
+                    let new_opp = StringDiffOp{
+                        kind : StringDiffOpKind::Substitute(
+                            s1.chars().nth(i).unwrap(),
+                            s2.chars().nth(i).unwrap() ),
+                        index : i as isize
+                    };
+                    opp_vec.push(new_opp)
+                }
+            }
+        }
+        return opp_vec;
+    }        
+
+    
+    fn distance<'a>(&self, s1: &'a str, s2: &'a str) -> usize{
+        let mut edit_distance = 0;
```

Comment:
This statement initializing the `edit_distance` variable should go after the error checking to avoid allocating memory whenever the caller hits that error. It should also specify the type via `0usize`